src/Orchard/ContentManagement/DefaultHqlQuery.cs

```diff
@@ -207,12 +207,18 @@ public class DefaultHqlQuery : IHqlQuery {
 
         public int Count() {
             ApplyHqlVersionOptionsRestrictions(_versionOptions);
-            var hql = ToHql(true);
-            hql = "select count(Id) from Orchard.ContentManagement.Records.ContentItemVersionRecord where Id in ( " + hql + " )";
-            return Convert.ToInt32(_session.CreateQuery(hql)
-                           .SetCacheable(true)
-                           .UniqueResult())
-                ;
+            var sql = ToSql(true);
+            sql = "SELECT count(*) as totalCount from (" + sql + ") t";
+            return Convert.ToInt32(_session.CreateSQLQuery(sql)
+                    .AddScalar("totalCount", NHibernateUtil.Int32)
+                    .UniqueResult());
+        }
+
+        public string ToSql(bool count) {
+            var sessionImp = (ISessionImplementor)_session;
+            var translatorFactory = new ASTQueryTranslatorFactory();
```

Comment:
To you need to have a new factory on every call:
- can it be per DefaultHqlQuery instance
- can it be static